prov/gni/src/gnix_mr.c

```diff
@@ -984,56 +1357,56 @@ static int __mr_cache_register(
 	key.length = length;
 
 	/* fastpath inuse */
-	ret = __mr_cache_lookup_inuse_fastpath(cache, &key, &entry);
+	ret = __mr_cache_search_inuse(cache, domain,
+			address, length, dst_cq_hndl, flags,
+			vmdh_index, &entry, &key);
 	if (ret == FI_SUCCESS)
 		goto success;
 
 	/* if we shouldn't introduce any new elements, return -FI_ENOSPC */
 	if (unlikely(cache->attr.hard_reg_limit > 0 &&
-			(atomic_get(&cache->inuse_elements) >=
+			(atomic_get(&cache->inuse.elements) >=
 					cache->attr.hard_reg_limit)))
-		return FI_ENOSPC;
+		return -FI_ENOSPC;
 
 	if (cache->attr.lazy_deregistration) {
 		/* if lazy deregistration is in use, we can check the
 		 *   stale tree
 		 */
-		ret = __mr_cache_lookup_stale_fastpath(cache, &key, &entry);
-		if (ret == FI_SUCCESS)
-			goto success;
-	}
-
-	/* slow path inuse */
-	ret = __mr_cache_lookup_inuse_slowpath(cache, &key, &entry);
-	if (ret == FI_SUCCESS)
-		goto success;
-
-	/* slow path stale */
-	if (cache->attr.lazy_deregistration) {
-		ret = __mr_cache_lookup_stale_slowpath(cache, &key, &entry);
-		if (ret == FI_SUCCESS)
+		ret = __mr_cache_search_stale(cache, domain,
+				address, length, dst_cq_hndl, flags,
+				vmdh_index, &entry, &key);
+		if (ret == FI_SUCCESS) {
+			cache->hits++;
 			goto success;
+		}
 	}
 
 	/* If the cache is full, then flush one of the stale entries to make
 	 *   room for the new entry. This works because we check above to see if
 	 *   the number of inuse entries exceeds the hard reg limit
 	 */
-	if ((atomic_get(&cache->inuse_elements) +
-			atomic_get(&cache->stale_elements)) == cache->attr.hard_reg_limit)
+	if ((atomic_get(&cache->inuse.elements) +
+			atomic_get(&cache->stale.elements)) == cache->attr.hard_reg_limit)
 		__mr_cache_flush(cache, 1);
 
-	ret = __mr_cache_create_registration(cache, mr, domain,
+	ret = __mr_cache_create_registration(cache, domain,
 			address, length, dst_cq_hndl, flags,
-			vmdh_index, mem_hndl, &entry);
+			vmdh_index, &entry, &key);
 	if (ret)
 		return ret;
 
+	cache->misses++;
+
 success:
-	mr->nic = entry->nic;
-	mr->key.address = entry->key.address;
-	mr->key.length = entry->key.length;
-	*mem_hndl = entry->mem_hndl;
+	entry->state = GNIX_CES_INUSE;
+	*mr = &entry->mr;
+
+	(*mr)->nic = entry->nic;
+	(*mr)->key.address = entry->key.address;
+	(*mr)->key.length = entry->key.length;
+	(*mr)->mem_hndl = entry->mem_hndl;
```

Comment:
It won't hurt anything since we would essentially be writing the same data
to the same fields. However, we could eliminate any strangeness in the
case that we find a cached entry and avoid those writes.
-- Jim




On 1/6/16, 12:53 PM, "Zach Tiffany" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_mr.c
><https://github.com/ofi-cray/libfabric-cray/pull/582#discussion_r48993541>
>:
>>  success:
>> -	mr->nic = entry->nic;
>> -	mr->key.address = entry->key.address;
>> -	mr->key.length = entry->key.length;
>> -	*mem_hndl = entry->mem_hndl;
>> +	entry->state = GNIX_CES_INUSE;
>> +	*mr = &entry->mr;
>> +
>> +	(*mr)->nic = entry->nic;
>> +	(*mr)->key.address = entry->key.address;
>> +	(*mr)->key.length = entry->key.length;
>> +	(*mr)->mem_hndl = entry->mem_hndl;
>Looks like you rewrite these fields in existing cache entries (found in
>one of the trees). Am I right? Is that ok? Just making sure.
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/582/files#r48993541>.
>
>